### PR TITLE
Fix annex B block function hoisting semantics

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes a regression introduced in https://github.com/babel/babel/pull/11801
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The regression was caught by test262: https://app.circleci.com/pipelines/github/babel/babel/4967/workflows/27043ada-f8d2-482e-810b-2593a17cf087/jobs/35381/tests

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12512"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-block-scoping/test/fixtures/general/issue-10046-async-generator/input.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// NOTE: When compiling async generators,
+// we run the async-generator-functions transform
+// as soon as in Program:enter, so when the block-scoping plugin
+// runs it has already been transpiled to a plain function.
+// The functions is thus visible to the outer scope.
+// This is a bug.
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-block-scoping/test/fixtures/general/issue-10046-async-generator/input.js:6]: Fixing this isn't easy and probably not really worth it since this a super edge case

**Comment by nicolo-ribaudo** [comment]
I tirggered the test262 run

**Review by existentialism — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]
